paasta_tools/kubernetes_tools.py

```diff
@@ -1865,13 +1868,90 @@ def update_stateful_set(
     )
 
 
+def get_events_for_object(
+    kube_client: KubeClient,
+    obj: Union[V1Pod, V1Deployment, V1StatefulSet, V1ReplicaSet],
+    kind: str,  # for some reason, obj.kind isn't populated when this function is called so we pass it in by hand
+) -> List[V1Event]:
+    host = KubeConfiguration().host
+
+    # The python kubernetes client doesn't support the V1Events API
+    # yet, so we have to make the request by hand (we need the V1Events
+    # API so that we can query by the involvedObject.name/kind)
+    events = kube_client.request(
+        "GET",
+        f"{host}/api/v1/namespaces/{obj.metadata.namespace}/events",
+        query_params={
+            "fieldSelector": f"involvedObject.name={obj.metadata.name},"
+            f"involvedObject.kind={kind}"
+        },
+    )
+    parsed_events = json.loads(events.data)
+    return parsed_events["items"]
+
+
+def get_all_events_for_service(
+    app: Union[V1Deployment, V1StatefulSet], kube_client: KubeClient
+) -> List:
+    """ There is no universal API for getting all the events pertaining to
+    a particular object and all its sub-objects, so here we just enumerate
+    all the kinds of objects that we care about, and get all the relevent
+    events for each of those kinds """
+    events = []
+    ls = (
+        f'paasta.yelp.com/service={app.metadata.labels["paasta.yelp.com/service"]},'
+        f'paasta.yelp.com/instance={app.metadata.labels["paasta.yelp.com/instance"]}'
+    )
+    for pod in kube_client.core.list_namespaced_pod(
+        app.metadata.namespace, label_selector=ls,
+    ).items:
+        events += get_events_for_object(kube_client, pod, "Pod")
+
+    for depl in kube_client.deployments.list_namespaced_deployment(
+        app.metadata.namespace, label_selector=ls,
+    ).items:
+        events += get_events_for_object(kube_client, depl, "Deployment")
+
+    for rs in kube_client.deployments.list_namespaced_replica_set(
+        app.metadata.namespace, label_selector=ls,
+    ).items:
+        events += get_events_for_object(kube_client, rs, "ReplicaSet")
+
+    for ss in kube_client.deployments.list_namespaced_stateful_set(
+        app.metadata.namespace, label_selector=ls,
+    ).items:
+        events += get_events_for_object(kube_client, ss, "StatefulSet")
+
+    return sorted(
+        events,
+        key=lambda x: (
+            x.get("lastTimestamp") or x.get("eventTime") or x.get("firstTimestamp")
```

Comment:
the assumption is that at least one of these exists, what if they don't? i'd add something like `or 0` in the end or fail explicitly with relevant error message.

Reply:
Fixed